### DISCUSSION TITLE
Getting Useragent in _app.tsx

### DISCUSSION THREAD

**Discussion by bwesner** [vercel/next.js#13614] [Help] [answered]
I want to show any user with a specific useragent (IE) a page that tells them to go download a supported browser, instead of letting them interact with the app.

I figured this logic belongs in `_app.tsx`. The problem is that some of my pages make use of HOCs like [withApollo](https://github.com/vercel/next.js/tree/canary/examples/with-apollo). So when I try to do something like this in `_app.tsx`:

```
import './styles.scss';

const App = ({ Component, pageProps, userAgent }: any): JSX.Element => {
    // do some check with userAgent here

    return <Component {...pageProps} />;
};

App.getInitialProps = async (ctx: any): Promise<any> => {
    const initialProps = await App.getInitialProps(ctx);
    const userAgent = typeof window === 'undefined' ? ctx.req.headers['user-agent'] : window.navigator.userAgent;

    return { userAgent, ...initialProps };
};

export default App;
```

I'm getting an infinite loop within the `getInitialProps` function (which makes sense). I've tried using https://www.npmjs.com/package/next-useragent too -- it gives me problems that seem to be related to having multiple HOCs chained together.

**Comment by jamesmosier** [comment]
```jsx
import App from 'next/app'

const MyApp = ({ Component, pageProps, userAgent }: any): JSX.Element => {
    // do some check with userAgent here

    return <Component {...pageProps} />;
};

MyApp.getInitialProps = async (ctx: any): Promise<any> => {
    const initialProps = await App.getInitialProps(ctx);
    const userAgent = typeof window === 'undefined' ? ctx.req.headers['user-agent'] : window.navigator.userAgent;

    return { userAgent, ...initialProps };
};

export default MyApp;
```

you'll want to import `App` from `next/app` and change the name of your component to something else (i.e. `MyApp`)



---

My 2 cents though would to just do all of this client side. There really isn't any benefit to user agent sniffing to display a message to IE users on the server.

And also as [the docs](https://nextjs.org/docs/advanced-features/custom-app) mention:

> [using getInitialProps in _app] This disables the ability to perform automatic static optimization, causing every page in your app to be server-side rendered.

**Reply by Vadorequest** [reply]
Definitely don't do that or you'll opt-out of static optimisation.

It's fine as an answer because the OP already uses getInitialProps in _app, but don't take this as "the right thing to do".

**Reply by bwesner** [reply]
Generally makes sense doing it client side only, thanks

**Comment by Vadorequest** [comment, marked as answer]
@bwesner I recommend you change your strategy, it won't work.

I've gone through the same issue, and here's what I've learned:
- You shouldn't rely on the Next.js app to display a message on deprecated/unsupported browser, because the code itself may crash on those browsers (lack of polyfills, etc.)
- Instead, the only way to properly display a message warning about the outdated browser is to load a JS script async, that doesn't use ES5+ features, so that it may work in all browsers (also, it won't slow down your app or increase bundle size, since it's async). The JS script should be hosted on a CDN (external to your Next.js app)
- A potential implementation for this is https://github.com/mikemaccana/outdated-browser-rework (demo at https://mikemaccana.github.io/outdated-browser-rework/)
	- But beware it won't play nicely with embedded browsers, such as Chrome from the facebook mobile app, see https://github.com/mikemaccana/outdated-browser-rework/issues/57 because it doesn't detect those are embedded browsers and still warn about it (and the end-user can't update those)

At this time, I don't have a working solution, I eventually disabled outdated-browser-rework because of above mentioned bug. But I hope this still helps figuring out the right way to solve your actual issue.

**Reply by bwesner** [reply]
Interesting, this makes sense, thanks. I will try to use that tool or perhaps some simplified adaptation of it.

**Comment by hardikmodi1** [comment]
We have taken different approach here. We have our custom server setup as described in the documentation here https://nextjs.org/docs/advanced-features/custom-server. So, we have added middleware in the server which uses browserslist and browserslist-useragentto determine whether userAgent is supported or not.

```
const userAgent = req.get('User-Agent');
const isBrowserSupported = UserAgent.matchesUA(userAgent, {
    browsers: pkg.browserslist.supported_browsers,
    ignoreMinor: true,
    ignorePatch: true,
    allowHigherVersions: true,
});
```
Based on this we can either redirect user to `/api/outdated` or just serve the html file from here.

Is this way seems fine or Is there a better way/place to detect browsers and redirect them to a static HTML page?
